src/mod_presence.erl

```diff
@@ -0,0 +1,613 @@
+    mongoose_acc:t().
+presence_broadcast_first(Acc0, FromJid, Packet, Presences, SocketSend) ->
+    Probe = presence_probe(),
+    _Acc1 = gb_sets:fold(fun(JID, Accum) ->
+                                ejabberd_router:route(FromJid, JID, Accum, Probe)
+                         end, Acc0, Presences#presences.pres_t),
+    case Presences#presences.pres_invis of
+        true ->
+            mongoose_c2s_acc:to_acc(Acc0, state_mod, {?MODULE, Presences});
+        false ->
+            {As, _Acc2} = gb_sets:fold(
+                               fun(JID, {A, Accum}) ->
+                                       Accum1 = ejabberd_router:route(FromJid, JID, Accum, Packet),
+                                       {gb_sets:add_element(JID, A), Accum1}
+                               end,
+                               {Presences#presences.pres_a, Acc0}, Presences#presences.pres_f),
+            NewPresences = Presences#presences{pres_a = As},
+            ToAcc = [{socket_send, SocketSend}, {state_mod, {?MODULE, NewPresences}}],
+            mongoose_c2s_acc:to_acc_many(Acc0, ToAcc)
+    end.
+
+-spec presence_probe() -> exml:element().
+presence_probe() ->
+    #xmlel{name = <<"presence">>,
+           attrs = [{<<"type">>, <<"probe">>}]}.
+
+-spec presence_unavailable_stanza() -> exml:element().
+presence_unavailable_stanza() ->
+    presence_unavailable_stanza(<<>>).
+
+-spec presence_unavailable_stanza(binary()) -> exml:element().
+presence_unavailable_stanza(<<>>) ->
+    #xmlel{name = <<"presence">>,
+           attrs = [{<<"type">>, <<"unavailable">>}]};
+presence_unavailable_stanza(Status) ->
+    StatusEl = #xmlel{name = <<"status">>,
+                      children = [#xmlcdata{content = Status}]},
+    #xmlel{name = <<"presence">>,
+           attrs = [{<<"type">>, <<"unavailable">>}],
+           children = [StatusEl]}.
+
+close_session_status(normal) ->
+    <<>>;
+close_session_status({shutdown, retries}) ->
+    <<"Too many attempts">>;
+close_session_status({shutdown, replaced}) ->
+    <<"Replaced by new connection">>;
+close_session_status(_) ->
+    <<"Unknown condition">>.
+
+-spec maybe_get_handler(mongoose_c2s:c2s_data()) -> presences().
+maybe_get_handler(StateData) ->
+    case mongoose_c2s:get_mod_state(StateData, ?MODULE) of
+        #presences{} = Presences -> Presences;
+        {error, not_found} -> #presences{}
+    end.
+
+-spec get_mod_state(mongoose_c2s:c2s_data()) -> presences() | {error, not_found}.
+get_mod_state(StateData) ->
+    mongoose_c2s:get_mod_state(StateData, ?MODULE).
+
+-spec get_priority_from_presence(exml:element()) -> priority().
+get_priority_from_presence(PresencePacket) ->
+    MaybePriority = exml_query:path(PresencePacket, [{element, <<"priority">>}, cdata], undefined),
+    case catch binary_to_integer(MaybePriority) of
+        P when is_integer(P), -128 =< P, P =< 128 -> P;
+        _ -> 0
+    end.
+
+-spec get_old_priority(presences()) -> priority().
```

Comment:
I don't like the name presences here, because it is anything but not `[#xmlel{name = <<"presence">>}]` (i.e. presences to me).

Maybe `pstate` could be better.

Reply:
It is correct that the name `presences` might suggest a list, but the type is clearly defined at the top, so I have no doubt what it means when reading the code. Anyway, `presence_state` might be better, but IMO it's ok as it is now.